[WIP] Added error codes from src/error.cpp into an enumeration

D4N opened the pull request:
As requested, I have put all error codes from src/error.cpp into an enumeration into error.hpp.

Please take a look whether they are reasonably named, I struggled a bit with coming up with good "abbreviations".

Replacing all `throw Error(X, ...)` with the appropriate value from the enum is going to take a while, thus please do not merge it yet. I'll probably write a script for that ;)

D4N committed:
Added error codes from src/error.cpp into an enumeration

clanmills commented:
Thanks for doing this.  Is it possible to "propagate" this enum throughout the code.  There's a table in src/error.cpp with the error numbers and a string.  Can the numbers be replaced with values from the enum?  And can all the "throws" in the code be converted from numbers to the symbolic name.

My other comment is about the names.  When you've update src/error.cpp, I will more easily review the enum "names" with the english strings.  Incidentally, because of the _N(macro), these string are localised by the i18n files in po/*.  I know nothing about i18n in Exiv2, however Thomsas (@tbeu) knows  something about that.  I don't thing the enum has any interaction with the po files.

I would prefer to see "camelCase" names being used in the code.  So, I would prepend ker to all the names.  k = konstant, er = Error Enumerator.  So the names would be kerSuccess and the like.  Perhaps you can ask Luis for his opinion about this.  I like to be able to look at a symbol and know it's derivation without thought.  So member variables end with _, enum values are something like kxyWhoFluDung.

Thanks very much for working on this.  The project is rolling along at amazing speed.  I've been unwell this week with a medical virus (flu symptoms).  This often happens to me on vacations.  I've found it rather difficult to think and even harder to work on anything during the last few days.  We're in Washington DC for the weekend and visiting ex-Adobe folks who now live here.  The weather is perfect.  30C and not a cloud in the sky.

D4N committed:
Incorporated error codes into errList

D4N commented:
I have changed the error constant names and included them into the `errList` in `src/error.cpp`. I'll try to write a simple script that will replace all the error numbers with the appropriate error codes. But the resulting diff is going to be huge, there are countless places where Errors are thrown.

That's very unfortunate that you got sick on your vacation (especially if it always happens). I hope you'll get well soon and can enjoy your trip to the fullest.

clanmills commented:
Thanks for doing this.  I'm not sure there are "countless" throws in the code.  I make it 677!
```
680 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2 $ grep throw src/*.cpp | wc
    677    4247   50067
681 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2
```
It's a big commit, however they'll be almost all one-line simple changes.  And we get the win that it's much easier to find where errors are being thrown.  Overall, I think this is a good change to make.

I'm hardly ever sick (at home or vacation).  It's my wife who has noticed this new pattern of "flu/cold" on vacation.  Anyway, she's done the driving while I've slept in the car.  We haven't changed our schedule or plans.  So, life is good.

piponazo commented:
This will be a great improvement for the code! Thanks for this contribution :)

@clanmills get well soon!

clanmills reviewed (approved):
This seems fine to me.  Can you explicitly assign kerSuccess = 0.  I think that's implied, however lets make it explit so we can say if ( !error ) to mean if error == kerSuccess.

D4N commented:
I have added the explicit `kerSuccess = 0` to `errors.hpp` and rebased the commits on top of the current head. Could you take a look at the enum names if you find them 'intuitive' enough? I struggled with some of them quite a bit.

Also please **don't merge** this yet. I would like to first settle on the enum names and then replace the numbers everywhere in the code.

clanmills commented on src/error.cpp:
```diff
@@ -50,67 +50,128 @@ namespace {
+          N_("Error %0: arg2=%2, arg3=%3, arg1=%1.") },
+        { Exiv2::kerSuccess,
+          N_("Success") },
+        { Exiv2::kerBriefError,
```
kerErrorMessage

clanmills reviewed (approved):
This is really good stuff.  Now that errors have names we can see there are duplicates.  Eg kerTooManyTiffDirectoryEntries (Andreas added?) and kerTiffDirectoryTooLarge (which I added)

I added "invalid type value detected in Image::printIFDStructure"  Can you change it to "invalid type in tiff structure" (more generic).

There's some inconsistency such as "Not valid XMP" and "Not a valid ICC Profile", let's have "Not valid ICC Profile".

kerCorruptedMetadata should be "Tiff Parsing Error".  It might be corrupted, or it could be something else!

Let's be consistent kerCapitol.  So kersetValueTooLarge -> kerSetValueTooLarge.  Possibly kerValueTooLarge.

There's a message inviting you to email Andreas.  Change that to "Report this (with your test data) to http://github.com/Exiv2/exiv2".

However it seems fine to me.  I don't think there's anything here that must be changed.  A little polish and we're done. The bigger task is to update all the throws to use the symbols.  I think you'll be able to write a sed script to do that for you.

It's really wonderful to see the issue count on github going down.  And I'm especially pleased that you and Luis and Michal have worked steadily without needing my time.

D4N commented:
I have changed the spelling of `kersetValueTooLarge`, `kerBriefError` and `kersetDataAreaValueTooLarge` to `kerValueTooLarge`, `kerErrorMessage` and kerDataAreaValueTooLarge`, respectively. The new commits have been pushed.

Concerning the merging/editing of some of the error messages, I've made a summary:

- change error message of `Exiv2::kerInvalidTypeValue` to "invalid type in tiff structure"
- merge `Exiv2::kerInvalidIccProfile` and `Exiv2::kerInvalidXMP` to `Exiv2::kerInvalidIccProfile`
- `kerCorruptedMetadata` to `kerTiffParsingError`, just a (stupid) question concerning this one: I believe one of the bugs I fixed where I have used this error were in the jpeg parser part, so how would this error message be applicable?
- merge `kerTooManyTiffDirectoryEntries` and `kerTiffDirectoryTooLarge`, I am actually unsure if they should be merged. `kerTooManyTiffDirectoryEntries` aka `Error 49` is used only once in the whole code in `src/tiffcomposite.cpp:1093` (which was indeed last edited by Andreas in 2010) however it is used in the context of writing too many Tiff directories to a file. `kerTiffDirectoryTooLarge` (aka `Error 55`) is actually only once used correctly (I assume) in `Image::printIFDStructure` (the other uses are in the network code and look like this: `Error(55, "Server", serverCode)`, so I assume that is wrong) where it reports that a Tiff directory is larger than `500`. In my opinion these are two different errors.

Since the merging of error messages will probably screw up the test suite real bad, I'd wait with that until I have rewritten it, so that it provides manageable diffs.

Also I think I'll open an issue concerning the odd usages of Error 55 in `src/basicio.cpp` and `samples/conntest.cpp`.

clanmills commented:
Most metadata is stored in a TIFF structure.  Even in a JPEG or PNG file!  Exif metadata is stored in an embedded tiff structure.  So, it's OK for Exiv2 to report a Tiff parsing error on a JPEG file.

XMP is XML embedded in the file.  IPTC is a binary structure which Exiv2 can decode and encode.  ICC is a highly structured entity, however Exiv2 treats it as a binary blob.  Exiv2 knows how to embed/extract/remove the ICC Profile in different file "containers", however Exiv2 has no knowledge structure of the ICC profile.

Error 55 Server is a "rogue".  I think Tuan added that for WebReady (performing IO using web protocols).  This is a good find.  Can you add a new error kerServerError.  It's probably 60 or something (although the numerical value is meaningless).

Can I ask you to see what happens if you throw an undefined error:  throw(911, "what happens next").  I hope throw handles this gracefully without crashing!

Good finds concerning kerTiffDirectoryTooLarge.  I added that a couple of years ago when I added printIFDStructure().  If a file has more than 500 entries in an IFD, I'm suspicious.  However it is arbitrary.

Don't touch Andreas' ingenious code src/tiff*.cpp without solid reason.  Changing **`throw`** to use enum ErrorCode is a solid reason.  Michal and I are working to consider how to refactor that code for BigTiff.

D4N committed:
Force error code usage to construct a Exiv2::BasicError

changed parameter code_'s type from int to ErrorCode => requires the
usage of the enum constants

D4N committed:
Replaced all hardcoded error codes with ker... constants

D4N commented:
I have replaced every single occurrence of hard-coded error codes in the code with the constants from the enumeration. However the process is not yet finished, as I have not yet applied the changes that @clanmills proposed. That's the next step, so please no merging yet.

For further reference, I have done most of the grind work with this Python script (unfortunately it is not aware of namespaces, it deletes trailing whitespaces and replaces windows line ends with unix ones, so some manual work is still required):
``` Python
# -*- coding: utf-8 -*-

import fileinput
import os
import re


ERROR_CODES = [
    "kerGeneralError",
    "kerSuccess",
    "kerErrorMessage",
    "kerCallFailed",
    "kerNotAnImage",
    "kerInvalidDataset",
    "kerInvalidRecord",
    "kerInvalidKey",
    "kerInvalidTag",
    "kerValueNotSet",
    "kerDataSourceOpenFailed",
    "kerFileOpenFailed",
    "kerFileContainsUnknownImageType",
    "kerMemoryContainsUnknownImageType",
    "kerUnsupportedImageType",
    "kerFailedToReadImageData",
    "kerNotAJpeg",
    "kerFailedToMapFileForReadWrite",
    "kerFileRenameFailed",
    "kerTransferFailed",
    "kerMemoryTransferFailed",
    "kerInputDataReadFailed",
    "kerImageWriteFailed",
    "kerNoImageInInputData",
    "kerInvalidIfdId",
    "kerValueTooLarge",
    "kerDataAreaValueTooLarge",
    "kerOffsetOutOfRange",
    "kerUnsupportedDataAreaOffsetType",
    "kerInvalidCharset",
    "kerUnsupportedDateFormat",
    "kerUnsupportedTimeFormat",
    "kerWritingImageFormatUnsupported",
    "kerInvalidSettingForImage",
    "kerNotACrwImage",
    "kerFunctionNotSupported",
    "kerNoNamespaceInfoForXmpPrefix",
    "kerNoPrefixForNamespace",
    "kerTooLargeJpegSegment",
    "kerUnhandledXmpdatum",
    "kerUnhandledXmpNode",
    "kerXMPToolkitError",
    "kerDecodeLangAltPropertyFailed",
    "kerDecodeLangAltQualifierFailed",
    "kerEncodeLangAltPropertyFailed",
    "kerPropertyNameIdentificationFailed",
    "kerSchemaNamespaceNotRegistered",
    "kerNoNamespaceForPrefix",
    "kerAliasesNotSupported",
    "kerInvalidXmpText",
    "kerTooManyTiffDirectoryEntries",
    "kerMultipleTiffArrayElementTagsInDirectory",
    "kerWrongTiffArrayElementTagType",
    "kerInvalidKeyXmpValue",
    "kerInvalidIccProfile",
    "kerInvalidXMP",
    "kerTiffDirectoryTooLarge",
    "kerInvalidTypeValue",
    "kerInvalidMalloc",
    "kerCorruptedMetadata",
    "kerArithmeticOverflow",
    "kerMallocFailed",
]

# kerMallocFailed is error 60 plus error 0 and error -1
assert(len(ERROR_CODES) == 60 + 1 + 1)


def replacer(match):
    err_code = int(match.group(1))
    return "Error(" + ERROR_CODES[err_code + 1]


def process_file(filename):
    for line in fileinput.input(files=(filename), inplace=True):
        print(re.sub('Error\(([0-9]+)', replacer, line.rstrip()))


for src_dir in ["src", "samples", "include"]:
    for root, _, files in os.walk(src_dir):
        for src_file in files:
            ext = os.path.splitext(src_file)[1]
            if ext in [".cpp", ".hpp"]:
                process_file(os.path.join(root, src_file))
```

clanmills commented:
This is really excellent work, Dan.  Thank you very much.

Two suggestion (more miracles:
1) I believe there are strings (`error_58_message`) in the test suite for those error messages.  Can this script generate the strings required by the test suite. 
2) Can  we use the same symbolic name in the test suite`kerCorruptedMetadata`

clanmills commented:
I've thought of something else.  Please preserve your python because I'm sure it'll be useful another day to make a global change to the code.

In svn days, we had svn://dev.exiv2.org/svn/team into which I stored various useful entities which are not part of the product.  And svn://dev.exiv2.org/svn/testdata for test files.  Perhaps we need a couple of additional repos exiv2/team and exiv2/testdata

piponazo commented on include/exiv2/error.hpp:
```diff
@@ -202,16 +270,20 @@ namespace Exiv2 {
         //@{
         //! Constructor taking only an error code
-        EXV_DLLLOCAL explicit BasicError(int code);
+        EXV_DLLLOCAL explicit BasicError(ErrorCodes code);
```
[suggestion] Normally the enum types do not have a plural name (`ErrorCodes`) but singular (`ErrorCode`). What we are passing to the function is a single error code, not a list of them. What do you think about changing the enum name ?

D4N replied:
Most of the enums in the source seem to use the singular for their name, so I'll change this for consistency. Thanks for bringing that up, it makes a lot more sense in the function signatures.

piponazo reviewed (approved):
Great work @D4N ! I just made a tiny suggestion, but of course this contribution is great to the project :+1:

piponazo commented:
Could this change go already into master or would we break something without applying the changes that @clanmills proposed ? 

I am always in favor of doing small iterations instead of merging huge PRs. I am afraid if you continue doing changes in this branch, it would be more difficult for the reviews to review your new changes.

clanmills commented:
There are a lot of lines of code (about 650) involved in this change.  However the changes are local and do not change the binary generated by the compiler.  We're only changing the token found by the compiler and not changing the code.  The code generated by the compiler is identical.
```
rmills@rmillsmbp:~/temp $ cat b.cpp 
#include <iostream>
enum Enum {
	Five = 5
};
int main(int, const char**)
{
	std::cout << 5 << std::endl;
	return 0;
}

rmills@rmillsmbp:~/temp $ make b  # compile << Five <<
c++     b.cpp   -o b
rmills@rmillsmbp:~/temp $ ls -alt b
-rwxr-xr-x  1 rmills  staff  9628 17 Feb 10:36 b
rmills@rmillsmbp:~/temp $ md5sum b
93be96eb6911c6a39801c2a3760d27f7  b
rmills@rmillsmbp:~/temp $ make b # compile << 5 <<
c++     b.cpp   -o b
rmills@rmillsmbp:~/temp $ md5sum b
93be96eb6911c6a39801c2a3760d27f7  b
rmills@rmillsmbp:~/temp $ 
```
The benefits of this change are:
1) Much improved code read-ability and clarity
2) We can grep the code to discover every use of an error message
3) We can use function signatures for improved code resilience

There are probably other benefits that have not occurred to me!

D4N committed:
Modified test suite so that case sensitive keys are possible

D4N committed:
Replace error variable names in test suite with enum error codes

D4N commented:
@piponazo Actually merging this now is a good idea, since once I start changing the error messages it'll break the testsuite again and will be delayed by #180. Nevertheless, this PR is already huge ;-) 

@clanmills Having additional repositories for scripts and test files is a good idea. We should probably discuss that in a separate issue, as we have already mentioned that idea a few times. I have also added some of the error codes to the test suite (that required a small change in the python code, hope that's fine in this PR).
There is also another advantage & disadvantage to using the enums as the error codes: you shouldn't be able to pass a wrong error code to `BasicError`, as the compiler will complain if you do this:
``` C++
throw Error(911);
```
However, it opens another can of worms: if you do something like this:
``` C++
int err_from_external_source = some_func();
if (err_from_external_source) { throw Error((ErrorCode)err_from_external_source); }
```
you are potentially invoking undefined behavior, if `err_from_external_source` is not a valid value of the enum `ErrorCode`. The issue here is, that we cannot really check in the BasicError constructor that `err_from_external_source` is valid, as compilers will simply optimize the check away (it's passed as a `ErrorCode` to the constructor, so the UB has already occurred before the constructor was invoked). However, the Error class is nowhere used in this way, so I think it should be fine.

Any objections to merging this in the current state?
